src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/StreamExtensions.cs

```diff
@@ -0,0 +1,140 @@
+    {
+        Debug.Assert(stream.CanWrite);
+        encoding ??= Encoding.UTF8;
+
+        var byteCount = encoding.GetMaxByteCount(text.Length);
+        using var _ = ArrayPool<byte>.Shared.GetPooledArray(byteCount, out var byteArray);
+
+#if NET
+        var usedBytes = encoding.GetBytes(text, byteArray);
+#else
+        var usedBytes = GetBytes(text, encoding, byteArray);
+#endif
+
+        WriteSize(stream, usedBytes);
+        return stream.WriteAsync(byteArray, 0, usedBytes, cancellationToken);
+
+#if !NET
+        static unsafe int GetBytes(string text, Encoding encoding, byte[] byteArray)
+        {
+            fixed (char* c = text)
+            fixed (byte* b = byteArray)
+            {
+                return encoding.GetBytes(c, text.Length, b, byteArray.Length);
+            }
+        }
+#endif
+    }
+
+    public static async Task<string> ReadStringAsync(this Stream stream, Encoding? encoding = null, CancellationToken cancellationToken = default)
+    {
+        Debug.Assert(stream.CanRead);
+        encoding ??= Encoding.UTF8;
+
+        var length = ReadSize(stream);
+
+        using var _ = ArrayPool<byte>.Shared.GetPooledArray(length, out var encodedBytes);
+        await stream.ReadAsync(encodedBytes, 0, length, cancellationToken).ConfigureAwait(false);
+        return encoding.GetString(encodedBytes, 0, length);
+    }
+
+    public static ProjectInfoAction ReadProjectInfoAction(this Stream stream)
+    {
+        var action = stream.ReadByte();
+        return action switch
+        {
+            0 => ProjectInfoAction.Update,
+            1 => ProjectInfoAction.Remove,
+            _ => throw Assumes.NotReachable()
+        };
+    }
+
+    public static void WriteProjectInfoAction(this Stream stream, ProjectInfoAction projectInfoAction)
+    {
+        stream.WriteByte(projectInfoAction switch
+        {
+            ProjectInfoAction.Update => 0,
+            ProjectInfoAction.Remove => 1,
+            _ => throw Assumes.NotReachable()
+        });
+    }
+
+    public static Task WriteProjectInfoRemovalAsync(this Stream stream, string intermediateOutputPath, CancellationToken cancellationToken)
+    {
+        WriteProjectInfoAction(stream, ProjectInfoAction.Remove);
+        return stream.WriteStringAsync(intermediateOutputPath, cancellationToken: cancellationToken);
+    }
+
+    public static Task<string> ReadProjectInfoRemovalAsync(this Stream stream, CancellationToken cancellationToken)
+    {
+        return stream.ReadStringAsync(cancellationToken: cancellationToken);
+    }
+
+    public static async Task WriteProjectInfoAsync(this Stream stream, RazorProjectInfo projectInfo, CancellationToken cancellationToken)
+    {
+        WriteProjectInfoAction(stream, ProjectInfoAction.Update);
+
+        var bytes = projectInfo.Serialize();
+        WriteSize(stream, bytes.Length);
+        await stream.WriteAsync(bytes, 0, bytes.Length, cancellationToken).ConfigureAwait(false);
```

Comment:
Can we not use the overload of `Serialize` that takes a stream? And similarly, the `Deserialize` overload, later?

Reply:
Serialize probably would be fine, but deserialize could potentially consume more data than it's supposed to? Not sure exactly...

Reply:
Looking at this we still need to write the size before writing the bytes, so unfortunately I think this is the best way.